### PR TITLE
feat: add MsgAddAuthorization to add or update and authorization and MsgRemoveAuthorization to remove a authorization

### PR DESCRIPTION
**Pull request by kingpinXD**
# Description
This pr adds the following messages, which can be used to update the authorization list 
- `MsgAddAuthorization` : Add or update an authorization 
- `MsgRemoveAuthorization`: Remove existing authorization

Closes: https://github.com/zeta-chain/node/issues/2288

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: x/authority/types/message_add_authorization.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+	return nil
+}
+
+func ValidateMsgURL(url string) error {
+	if len(url) == 0 {
+		return errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "message URL cannot be empty")
+	}
+	return nil
+}
```
Comment (kingpinXD) [inline x/authority/types/message_add_authorization.go:65]: It is keeping the validation logic simple, as this is an admin message. 
We can add stricter logic such as as the following for example 

```
vals := strings.Split(url, ".")
if vals[0] != "/zetachain" {
    return errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "message URL must start with '/zetachain'")
}
if vals[1] != "zetacore" {
    return errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "message URL must start with '/zetachain.zetacore'")
}
if len(vals) != 4 {
    return errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "message URL must contain 5 parsable parts")
}

``` 

Imo it seems unnecessary , but would love to hear some thoughts

File: x/authority/keeper/msg_server_add_authorization.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	authorizationList.SetAuthorization(types.Authorization{MsgUrl: msg.MsgUrl, AuthorizedPolicy: msg.AuthorizedPolicy})
+
+	// validate the authorization list after adding the authorization as a precautionary measure.
+	err := authorizationList.Validate()
+	if err != nil {
+		return nil, errorsmod.Wrap(err, "authorization list is invalid")
+	}
+
+	k.SetAuthorizationList(ctx, authorizationList)
+	return &types.MsgAddAuthorizationResponse{}, nil
```
Comment (kingpinXD) [inline x/authority/keeper/msg_server_add_authorization.go:37]: Choose not to emit an event , as this message would be submitted by the policy owners and might not need tobe tracked via an event

File: x/authority/types/message_add_authorization.go (hunk 1 above)
Comment (lumtis) [reply x/authority/types/message_add_authorization.go:65]: I think keeping it generic is good.

I think this is how it is in the `authz` module

File: x/authority/keeper/msg_server_add_authorization_test.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+		msg := &types.MsgAddAuthorization{
+			Creator:          admin,
+			MsgUrl:           url,
+			AuthorizedPolicy: types.PolicyType_groupAdmin,
+		}
+
+		_, err := msgServer.AddAuthorization(sdk.WrapSDKContext(ctx), msg)
+		require.NoError(t, err)
+
+		authorizationList, found := k.GetAuthorizationList(ctx)
```
Comment (lumtis) [inline x/authority/keeper/msg_server_add_authorization_test.go:152]: Maybe we can check previous and new list len are the same

File: x/authority/keeper/msg_server_add_authorization_test.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+		policy, err := authorizationList.GetAuthorizedPolicy(url)
+		require.NoError(t, err)
+		require.Equal(t, types.PolicyType_groupAdmin, policy)
+	})
+
+	t.Run("successfully set authorization when list is not found ", func(t *testing.T) {
+		k, ctx := keepertest.AuthorityKeeper(t)
+		admin := keepertest.SetAdminPolices(ctx, k)
+		msgServer := keeper.NewMsgServerImpl(*k)
+		url := "/zetachain.zetacore.sample.ABC"
```
Comment (lumtis) [inline x/authority/keeper/msg_server_add_authorization_test.go:112]: Looks like this variable could be set as a const in the file

File: x/authority/keeper/msg_server_add_authorization_test.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+			AuthorizedPolicy: types.PolicyType_groupAdmin,
+		}
+
+		_, err := msgServer.AddAuthorization(sdk.WrapSDKContext(ctx), msg)
+		require.NoError(t, err)
+
+		authorizationList, found := k.GetAuthorizationList(ctx)
+		require.True(t, found)
+		policy, err := authorizationList.GetAuthorizedPolicy(url)
+		require.NoError(t, err)
```
Comment (lumtis) [inline x/authority/keeper/msg_server_add_authorization_test.go:35]: We can check the list length is incremented by 1

File: x/authority/keeper/msg_server_remove_authorization_test.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+		authorizationList, found := k.GetAuthorizationList(ctx)
+		require.True(t, found)
+		_, err := authorizationList.GetAuthorizedPolicy(url)
+		require.NoError(t, err, types.ErrAuthorizationNotFound)
+
+		_, err = msgServer.RemoveAuthorization(sdk.WrapSDKContext(ctx), msg)
+		require.NoError(t, err)
+
+		authorizationList, found = k.GetAuthorizationList(ctx)
```
Comment (lumtis) [inline x/authority/keeper/msg_server_remove_authorization_test.go:36]: We can check the list length is decremented

File: x/authority/types/policytype.go [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package types
```
Comment (lumtis) [inline x/authority/types/policytype.go:1]: File to be named `policy_type`

File: x/authority/types/policytype_test.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
+package types_test
```
Comment (lumtis) [inline x/authority/types/policytype_test.go:1]: Same

File: x/authority/types/policytype_test.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+			wantErr:    false,
+		},
+		{
+			name:       "valid groupAdmin",
+			policyType: types.PolicyType_groupAdmin,
+			wantErr:    false,
+		},
+		{
+			name:       "invalid policy type",
+			policyType: types.PolicyType(20),
```
Comment (lumtis) [inline x/authority/types/policytype_test.go:34]: We should check `max+1` for edge case
We can use the len of map object generated by proto

File: x/authority/keeper/msg_server_add_authorization.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+	if !k.IsAuthorized(ctx, msg.Creator, types.PolicyType_groupAdmin) {
+		return nil, errorsmod.Wrap(
+			types.ErrUnauthorized,
+			"AddAuthorization can only be executed by the admin policy account",
+		)
+	}
+
+	authorizationList, _ := k.GetAuthorizationList(ctx)
+	authorizationList.SetAuthorization(types.Authorization{MsgUrl: msg.MsgUrl, AuthorizedPolicy: msg.AuthorizedPolicy})
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/authority/keeper/msg_server_add_authorization.go]: bcff29fa9b12893e7c7e65ce855fc08d20e256de

File: typescript/zetachain/zetacore/authority/policies_pb.d.ts [hunk 11]
```diff
@@ -34,6 +34,15 @@ export declare enum PolicyType {
    * @generated from enum value: groupAdmin = 2;
    */
   groupAdmin = 2,
+
+  /**
+   * protocol parameters or moving funds
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply typescript/zetachain/zetacore/authority/policies_pb.d.ts:39]: I think it is carried over from groupAdmin due to the multiline comments

File: x/authority/client/cli/tx_add_authorizations_test.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+	"github.com/zeta-chain/zetacore/x/authority/client/cli"
+	"github.com/zeta-chain/zetacore/x/authority/types"
+)
+
+func Test_GetPolicyType(t *testing.T) {
+	tt := []struct {
+		name                 string
+		policyTypeString     string
+		expecectedPolicyType types.PolicyType
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/authority/client/cli/tx_add_authorizations_test.go]: a49e5def25b2a7b21d792848b3734c4a315f9a08

File: x/authority/client/cli/tx.go [hunk 13]
```diff
@@ -22,6 +22,8 @@ func GetTxCmd() *cobra.Command {
 	cmd.AddCommand(
 		CmdUpdatePolices(),
```
Comment (lumtis) [inline x/authority/client/cli/tx.go:23]: Typo -> `CmdUpdatePolicies`

File: x/authority/keeper/msg_server_add_authorization_test.go [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+	"github.com/zeta-chain/zetacore/testutil/sample"
+	"github.com/zeta-chain/zetacore/x/authority/keeper"
+	"github.com/zeta-chain/zetacore/x/authority/types"
+)
+
+func TestMsgServer_AddAuthorization(t *testing.T) {
+	const url = "/zetachain.zetacore.sample.ABC"
+	t.Run("successfully add authorization of type admin to existing authorization list", func(t *testing.T) {
+		k, ctx := keepertest.AuthorityKeeper(t)
+		admin := keepertest.SetAdminPolices(ctx, k)
```
Comment (lumtis) [inline x/authority/keeper/msg_server_add_authorization_test.go:19]: Typo -> `SetAdminPolicies`

**Review by lumtis — COMMENTED** [review]
Looks good, reporting some typos

**Review by lumtis — APPROVED** [review]

**Pull request merged**